dbms/src/Flash/Planner/plans/PhysicalExpand.cpp

```diff
@@ -0,0 +1,117 @@
+    const String & executor_id,
+    const LoggerPtr & log,
+    const tipb::Expand & expand,
+    const PhysicalPlanNodePtr & child)
+{
+    assert(child);
+
+    child->finalize();
+
+    if (unlikely(expand.grouping_sets().empty()))
+    {
+        //should not reach here
+        throw TiFlashException("Expand executor without grouping sets", Errors::Planner::BadRequest);
+    }
+
+    DAGExpressionAnalyzer analyzer{child->getSchema(), context};
+    ExpressionActionsPtr before_expand_actions = PhysicalPlanHelper::newActions(child->getSampleBlock(), context);
+
+
+    auto shared_expand = analyzer.buildExpandGroupingColumns(expand, before_expand_actions);
+
+    // construct sample block.
+    NamesAndTypes expand_output_columns;
+    auto child_header = child->getSchema();
+    for (const auto & one : child_header)
+    {
+        expand_output_columns.emplace_back(one.name, shared_expand->isInGroupSetColumn(one.name) ? makeNullable(one.type) : one.type);
+    }
+    expand_output_columns.emplace_back(shared_expand->grouping_identifier_column_name, shared_expand->grouping_identifier_column_type);
+
+    auto physical_expand = std::make_shared<PhysicalExpand>(
+        executor_id,
+        expand_output_columns,
+        log->identifier(),
+        child,
+        shared_expand,
+        Block(expand_output_columns));
+
+    return physical_expand;
+}
+
+
+void PhysicalExpand::expandTransform(DAGPipeline & child_pipeline, Context & context)
+{
+    auto expand_actions = PhysicalPlanHelper::newActions(child_pipeline.firstStream()->getHeader(), context);
+    expand_actions->add(ExpressionAction::expandSource(shared_expand));
+    String expand_extra_info = fmt::format("expand, expand_executor_id = {}", execId());
+    child_pipeline.transform([&](auto & stream) {
+        stream = std::make_shared<ExpressionBlockInputStream>(stream, expand_actions, log->identifier());
+        stream->setExtraInfo(expand_extra_info);
+    });
+}
+
+void PhysicalExpand::transformImpl(DAGPipeline & pipeline, Context & context, size_t max_streams)
+{
+    child->transform(pipeline, context, max_streams);
+    expandTransform(pipeline, context);
+}
+
+void PhysicalExpand::finalize(const Names & parent_require)
+{
+    FinalizeHelper::checkSchemaContainsParentRequire(schema, parent_require);
+    Names required_output;
+    required_output.reserve(shared_expand->getGroupSetNum()); // grouping set column should be existed in the child output schema.
+    auto name_set = std::set<String>();
+    shared_expand->getAllGroupSetColumnNames(name_set);
+    // append parent_require column it may expect self-filled groupingID.
+    for (const auto & one : parent_require)
+    {
+        if (one != Expand::grouping_identifier_column_name)
+        {
+            name_set.insert(one);
+        }
+    }
+    for (const auto & grouping_name : name_set)
+    {
+        required_output.emplace_back(grouping_name);
+    }
+    child->finalize(required_output);
```

Comment:
Because `child->finalize()` has been called in `PhysicalExpand::build`
```suggestion
```

Reply:
maybe we should remove the child.finalize() in build func. (logic steps here are necessary from trials and errors)

Reply:
yes, IMO, the `child.finalize()` in build func should be removed.
`finalize` should only be called at once.